src/ray/pubsub/subscriber.cc

```diff
@@ -22,37 +22,59 @@ namespace pubsub {
 /// SubscriberChannel
 ///////////////////////////////////////////////////////////////////////////////
 
-void SubscriberChannel::Subscribe(
-    const rpc::Address &publisher_address, const std::string &key_id,
-    SubscriptionCallback subscription_callback,
+bool SubscriberChannel::Subscribe(
+    const rpc::Address &publisher_address, const std::optional<std::string> &key_id,
+    SubscriptionItemCallback subscription_callback,
     SubscriptionFailureCallback subscription_failure_callback) {
   cum_subscribe_requests_++;
   const auto publisher_id = PublisherID::FromBinary(publisher_address.worker_id());
 
-  RAY_CHECK(subscription_map_[publisher_id]
-                .subscription_callback_map
-                .emplace(key_id, std::make_pair(std::move(subscription_callback),
-                                                std::move(subscription_failure_callback)))
-                .second);
+  if (key_id) {
+    return subscription_map_[publisher_id]
+        .per_entity_subscription
+        .try_emplace(*key_id, std::make_pair(std::move(subscription_callback),
+                                             std::move(subscription_failure_callback)))
+        .second;
+  }
+  auto &all_entities_subscription =
+      subscription_map_[publisher_id].all_entities_subscription;
+  if (all_entities_subscription != nullptr) {
+    return false;
+  }
```

Comment:
Does this mean if we subscribe twice it won't take any effect?
What if there are two places subscribing to the same node? Should this be a vector? 
Or should we CHECK to make sure it's not happening like before?

Reply:
Currently the second subscription will be a no-op, and Subscribe() / SubscribeAll() will return false. The existing redis client API does not support multiple subscriptions of the same channel / ID from the same subscriber either. If we want to support multiple subscriptions to the same channel & ID, we would need to add a subscription handle which is non-trivial refactoring.
Ideally a new subscription can cancel the previous subscription, but the GRPC wrapper Ray uses does not expose cancel. Checking failing is a bit brittle since this is not a irrecoverable error. Letting caller decide how to handle this seems better.

Reply:
synced offline. could you add nodiscard annotation for these method to make sure the compiler can catch the error?

Reply:
Added [[nodiscard]] to `Subscribe()` and `SubscribeChannel()` functions.